### PR TITLE
Trap Refactor

### LINKED ISSUE #7219: When multiple people are trapped in the large trap, using release on the trap only frees the most recent person and bugs the others into being stuck to the floor
Checklist before you submit an issue! Feel free to partially or fully delete this in your eventual report.
* Please be specific in your description of the issue. Explain the desired result (what should be happening) and the actual result (what is happening). Simply stating that [something happens] is not very helpful at describing what's wrong.
When at least two people are in a trap, and a third enters, when someone else comes along and releases them using the verb "Release Animal," only the most recent person is released, and the others are stuck to the floor forever. Ideally it should unstick all that were in the trap. I believe this would happen even if there were only two in the trap and a third person released. The first person trapped would be stuck. Haven't witnessed firsthand, however.
* Have you tried reproducing the issue? If yes, steps to reproduce it would help immensely!
This has happened to me twice now.
Both times I was the third person in the trap and a fourth person came to release us.

The biggest question is, should multiple people/mobs be allowed to fall into the same large trap? Preventing multiple mobs from entering de facto fixes this. If not, captured needs to be a list.

### LINKED ISSUE #7292: Trap has no distance checks and is exceptionally buggy.
Trap "sneaking" has no distance checks. You can teleport towards the trap if you click and drag the trap sprite on yourself.

Also you're forced inside the actual contents of the trap if you click and drag the trap to yourself, but fail to sneak passed the trap. You can't get out this way unless you do it again.

### PR DESCRIPTION
**Pull request by JohnWildkins**
it's too late in the year to call it spring cleaning

tweaks and fixes:
- large traps must now be anchored and then set (deployed) before use, with a wrench and open hand respectively
- you can no longer pass through traps that have caught something
- you can still buckle things into large traps without them being anchored or deployed
- fix not being able to anchor large traps to the floor
- fix some forms of capture not capturing properly and some forms of release not releasing properly
- incorporates more or less everything from #7230 incl. fixes #7219 
- traps give engineering instead of material research when analyzed
- you can no longer teleport across the map by sneaking into traps, fixes #7292 
- things that can't be captured by a trap no longer trigger the trap

backend:
- replaces a lot of similar behavior with defined procs
- unifies the meaning of "deployed", i.e. ready to capture
- everything now works off of buckling in one way or another
- intercept MouseDrop_T so you can't buckle at weird times anymore
- cleaned up logic in places
- flipped icon states around to match deployed state

to-do:
- [ ] test everything more

remaining issues:
- you can't catch spiderlings anymore, although honestly i'm not sure you ever could, because they're obj/effects and you can't buckle those; let us pray that matt's refactor fixes this

### CODE REVIEW

**Comment by JohnWildkins** [comment]
!wip

**Comment by VTCobaltblood** [comment]
this guy should be head dev instead of arrow768

**Comment by Sindorman** [comment]
> you can no longer pass through traps that have caught something

I am not a fan of that, because it was a requested feature by the lore dev, when trap was originally developed

**Comment by JohnWildkins** [comment]
> I am not a fan of that, because it was a requested feature by the lore dev, when trap was originally developed

That should say _large_ traps, because I honestly don't know you would be able to pass through a cage capable of holding a human being while it's actively holding one. That being said, it's not a hill I need to die on, and it's easily revertable if that's what the original request was.

> Traps were not intended to have prey buckled to them as like chairs. That is a bug side-effect that should be fixed and is fixed in #7230

I think it's a quality-of-life improvement. They're as much cages as traps, and in such a way allowing for buckling under certain conditions is no different than what you'd need to do otherwise to accomplish the same effect, which is basically harm-intent shoving them into the trap or grab-throwing them. I'll probably add an equivalent timer to it as the other options at the very least, though.

**Comment by JohnWildkins** [comment]
okay it's definitely not perfect but I need some second and third eyes on this and ideally someone else to test and make sure I didn't break anything

**Comment by JohnWildkins** [comment]
!review

File: code/game/objects/items/weapons/traps.dm [hunk 1]
```diff
@@ -395,145 +378,85 @@
...
+			usr.forceMove(loc)
+			usr.visible_message("<span class='notice'>[usr] passes through \the [src] without triggering it.</span>",
+							"<span class='notice'>You pass through \the [src] without triggering it.</span>"
+			)
+		else
+			usr.forceMove(loc)
+			usr.visible_message("<span class='warning'>[usr] accidentally triggered \the [src]!</span>",
+							"<span class='warning'>You accidentally trigger \the [src]!</span>"
+			)
+			capture(usr)
```
Comment (skull132) [inline code/game/objects/items/weapons/traps.dm]: `user` over `usr`.

File: code/game/objects/items/weapons/traps.dm [hunk 2]
```diff
@@ -11,36 +11,41 @@
...
+		"<span class='danger'>You begin deploying \the [src]!</span>",
+		"You hear the slow creaking of a spring."
+		)
+
+	if (do_after(user, 5 SECONDS))
 		user.visible_message(
-			"<span class='danger'>[user] starts to deploy \the [src].</span>",
-			"<span class='danger'>You begin deploying \the [src]!</span>",
-			"You hear the slow creaking of a spring."
+			"<span class='danger'>[user] has deployed \the [src].</span>",
```
Comment (NonQueueingMatt) [inline code/game/objects/items/weapons/traps.dm]: This should be present simple, like every other action in the game

File: code/game/objects/items/weapons/traps.dm [hunk 3]
```diff
@@ -11,36 +11,41 @@
...
+			"<span class='danger'>You have deployed \the [src]!</span>",
```
Comment (NonQueueingMatt) [inline code/game/objects/items/weapons/traps.dm]: This should be present simple, like every other action in the game

File: code/game/objects/items/weapons/traps.dm [hunk 4]
```diff
@@ -110,7 +116,7 @@
 		bear.anger += 15//traps make bears really angry
 		bear.instant_aggro()
 
-/obj/item/weapon/trap/Crossed(AM as mob|obj)
+/obj/item/weapon/trap/Crossed(AM)
```
Comment (NonQueueingMatt) [inline code/game/objects/items/weapons/traps.dm]: shouldnt this be atom/movable/AM?

File: code/game/objects/items/weapons/traps.dm [hunk 5]
```diff
@@ -242,25 +248,19 @@
 
 	breakout = FALSE
 	to_chat(escapee, "<span class='warning'>You successfully break out!</span>")
-	visible_message("<span class='danger'>\the [escapee] successfully broke out of \the [src]!</span>")
+	visible_message("<span class='danger'>\The [escapee] successfully broke out of \the [src]!</span>")
```
Comment (NonQueueingMatt) [inline code/game/objects/items/weapons/traps.dm]:
```suggestion
	visible_message("<span class='danger'>\The [escapee] successfully breaks out of \the [src]!</span>")
```

File: code/game/objects/items/weapons/traps.dm [hunk 6]
```diff
@@ -349,41 +340,33 @@
...
-	if(istype(W, /obj/item/weapon/reagent_containers) && contents.len)
-		var/mob/living/L = pick(contents)
-		W.afterattack(L, user, TRUE)
-	else if(W.iswelder())
+/obj/item/weapon/trap/animal/attackby(obj/item/W, mob/user)
+	if(W.iswelder())
 		var/obj/item/weapon/weldingtool/WT = W
-		if (!WT.welding)
+		if(!WT.welding)
 			to_chat(user, "Your \the [W] is off!")
```
Comment (NonQueueingMatt) [inline code/game/objects/items/weapons/traps.dm]: This could use a span class

File: code/game/objects/items/weapons/traps.dm [hunk 7]
```diff
@@ -349,41 +340,33 @@
...
-		if(WT.remove_fuel(0, user))
-			user.visible_message("<span class='notice'>[user] is sliced \the [src]!</span>",
-								 "You sliced \the [src]!")
-			new /obj/item/stack/rods(src.loc, resources["rods"])
-			if(resources.len == 2)
-				new /obj/item/stack/material/steel(src.loc, resources["metal"])
-			release(user)
-			qdel(src)
+		if (do_after(user, 30/W.toolspeed, act_target = src))
+			if(WT.remove_fuel(0, user))
```
Comment (NonQueueingMatt) [inline code/game/objects/items/weapons/traps.dm]: This removes... zero fuel?

File: code/game/objects/items/weapons/traps.dm [hunk 8]
```diff
@@ -349,41 +340,33 @@
...
+				user.visible_message("<span class='notice'>[user] is sliced \the [src]!</span>",
```
Comment (NonQueueingMatt) [inline code/game/objects/items/weapons/traps.dm]:
```suggestion
				user.visible_message("<span class='notice'>[user] slices \the [src] open!</span>",
```

File: code/game/objects/items/weapons/traps.dm [hunk 9]
```diff
@@ -349,41 +340,33 @@
...
+									"You sliced \the [src]!")
```
Comment (NonQueueingMatt) [inline code/game/objects/items/weapons/traps.dm]:
```suggestion
									"You slice \the [src] open!")
```
Also pretty sure this one needs a span class.

File: code/game/objects/items/weapons/traps.dm [hunk 10]
```diff
@@ -395,145 +378,85 @@
...
+	)
+	if(do_after(user, 2 SECONDS, act_target = src))
+		if(prob(pct))
+			user.forceMove(loc)
+			user.visible_message("<span class='notice'>[user] passes through \the [src] without triggering it.</span>",
+							"<span class='notice'>You pass through \the [src] without triggering it.</span>"
+			)
+		else
+			user.forceMove(loc)
+			user.visible_message("<span class='warning'>[user] accidentally triggered \the [src]!</span>",
```
Comment (NonQueueingMatt) [inline code/game/objects/items/weapons/traps.dm]:
```suggestion
			user.visible_message("<span class='warning'>[user] accidentally triggers \the [src]!</span>",
```

File: code/game/objects/items/weapons/traps.dm [hunk 11]
```diff
@@ -607,55 +523,66 @@
 						/mob/living/simple_animal/hostile/carp, /mob/living/simple_animal/hostile/bear, /mob/living/simple_animal/hostile/alien, /mob/living/simple_animal/hostile/giant_spider,
 						/mob/living/simple_animal/hostile/commanded/dog, /mob/living/simple_animal/hostile/retaliate/cavern_dweller, /mob/living/carbon/human/)
 
-/obj/item/weapon/trap/animal/large/attack_hand(mob/user as mob)
-	return
+/obj/item/weapon/trap/animal/large/attack_hand(mob/user)
+	if(user.loc == loc)
+		to_chat(user, span("warning", "You can't do that from the inside!"))
```
Comment (NonQueueingMatt) [inline code/game/objects/items/weapons/traps.dm]: Can't you go onto the trap's tile without triggering it? If so, this isn't a good check.

**Pull request merged**